### ISSUE TITLE
Error when opening a synced Realm with an expired access token

### ISSUE THREAD

**Issue by ivanBereznev** [realm/realm-js#3667]
## Goals

Trying to open a synced realm.

## Expected Results

If session is expired I expect a promise rejection that can be handled in a proper way.

## Actual Results

Instead I have this error in log:
```
Connection[1]: Session[1]: client_reset_config = false, Realm exists = true, async open = false, client reset = false
Connection[1]: Connected to endpoint '52.49.130.120:443' (from '192.168.1.41:59359')
Connection[1]: Verifying server SSL certificate using 155 root certificates
ERROR: Connection[1]: Websocket: Expected HTTP response 101 Switching Protocols, but received:
HTTP/1.1 401 Unauthorized
cache-control: no-cache, no-store, must-revalidate
connection: close
content-length: 190
content-type: application/json
date: Sat, 27 Mar 2021 21:05:00 GMT
server: envoy
vary: Origin
x-envoy-max-retries: 0
x-frame-options: DENY


Connection[1]: Connection closed due to error
```

## Steps to Reproduce

Open a realm with an expired auth.

## Code Sample
```
const app = new Realm.App(MY_APP_ID)
const user = app.currentUser
if (!user) {
  return
}

let realm
try {
  realm = await Realm.open({
    schema: [....],
    sync: {
      error: () => { /* ... logging errors here... */},
      user: user,
      partitionValue: 'my-partition-value'
    }
  })
} catch (e) {
  // I expect to handle error here
}
```

## Version of Realm and Tooling

- Realm JS SDK Version: 10.2.0
- Node or React Native: Node 15.11.0
- Client OS & Version: Debian GNU/Linux 9.13
- Which debugger for React Native: ?/None

**Comment by kneth** [comment]
Two questions:

1. How to you log in your user? 
2. What is your `sync` configuration?

**Comment by ivanBereznev** [comment]
1. When there is no current user, I log in like this:
```
await app.logIn(Realm.Credentials.emailPassword('email@example.com', 'password'))
```
When I log in and then stop my app, on next start `app.currentUser` is usually here, but it may have an expired session, and if this is the case, then I get the described error when opening a realm.
2. Sorry, missed it in the original description, it looks like this:
```
realm = await Realm.open({
    schema: [....],
    sync: {
      error: () => { /* ... logging errors here... */},
      user: user,
      partitionValue: 'my-partition-value'
    }
  })
```

**Comment by kneth** [comment]
> and then stop my app, on next start

How long a pause do you have? If it is too long, sync might require to be [reenabled](https://docs.mongodb.com/realm/reference/terminating-and-reenabling-realm-sync/).

**Comment by ivanBereznev** [comment]
> How long a pause do you have? If it is too long, sync might require to be reenabled.
Not long, maybe about 30-40 minutes, I suppose it's just big enough to have access tokens outdated or something like that, because the other way to reproduce the error is to revoke all user sessions using MongoDB Realm UI and then start app once again.

**Comment by kneth** [comment]
The access token will expire in 30 minutes. I do observe a similar behaviour as you describe, but unfortunately I haven't been able to find the root cause. `app.currentUser` is a valid object, and `app.currentUser.state` will be `LoggedIn` even after 30 minutes.

I would imagine the token is refreshed if you do `await app.currentUser.refreshCustomData()`.

**Comment by nirinchev** [comment]
The issue is that OS doesn't handle failure to upgrade the websocket connection. See https://jira.mongodb.org/browse/RCORE-473 (private issue) for reference.

**Comment by kneth** [comment]
The fix was released in v10.4.1. I am closing the issue, and please reopen if you still experience it.
